### PR TITLE
Added RPYT Setpoint

### PR DESCRIPTION
**Pull request by Pedro-Roque**
Setpoint for Roll, Pitch, Yawrate, Thrust added, from ethz-asl fork.

### CODE REVIEW

**Comment by Pedro-Roque** [comment]
@TSC21 good to go?

**Comment by mhkabir** [comment]
Looks nice! Can you guys add some documentation on how to compute the thrust scaling factor and the yaw scaling factor?

**Comment by mhkabir** [comment]
Also - why do we need a scaler for yawrate? PX4 should accept yawrate setpoints in rad/s and normalised thrust setpoints [0-1].

File: mavros/src/plugins/setpoint_raw.cpp [hunk 1]
```diff
@@ -45,9 +47,26 @@ class SetpointRawPlugin : public plugin::PluginBase,
 
 		bool tf_listen;
 
+		ignore_rpyt_messages_ = false;
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: init this on the constructor instead: https://github.com/mavlink/mavros/pull/964/files#diff-146253845b95f1e30b7d5bd15e269768R41

File: mavros/src/plugins/setpoint_raw.cpp [hunk 2]
```diff
@@ -68,8 +87,10 @@ class SetpointRawPlugin : public plugin::PluginBase,
 	friend class SetAttitudeTargetMixin;
 	ros::NodeHandle sp_nh;
 
-	ros::Subscriber local_sub, global_sub, attitude_sub;
+	ros::Subscriber local_sub, global_sub, attitude_sub, rpyt_sub;
 	ros::Publisher target_local_pub, target_global_pub, target_attitude_pub;
+	double thrust_scaling_, system_mass_kg_, yaw_rate_scaling_;
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: Please remove the last underscore of each variable.

File: mavros/src/plugins/setpoint_raw.cpp [hunk 3]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+			ROS_FATAL("Recieved roll_pitch_yaw_thrust_rate message, but ignore_rpyt_messages_ is true: "
+	 	        "the most likely cause of this is a failure to specify the thrust_scaling_factor,	"
+			"yaw_rate_scaling_factor or system_mass_kg parameters");
+		     	return;
+		}
+
+		//the masks are much more limited than the docs would suggest so we don't use them
+		uint8_t type_mask = 0;
+
+		geometry_msgs::Quaternion orientation = tf::createQuaternionMsgFromRollPitchYaw( msg->roll, msg->pitch, 0);
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: `geometry_msgs::Quaternion orientation = tf::createQuaternionMsgFromRollPitchYaw(msg->roll, msg->pitch, 0);`

File: mavros/src/plugins/setpoint_raw.cpp [hunk 4]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+		
+		double thrust = std::min(1.0, std::max(0.0, msg->thrust.z * thrust_scaling_ * system_mass_kg_));
+		Eigen::Quaterniond desired_orientation;
+		Eigen::Vector3d body_rate;
+		tf::quaternionMsgToEigen(orientation, desired_orientation);
+		
+		// Transform desired orientation to represent aircraft->NED,
+		// MAVROS operates on orientation of base_link->ENU
+		
+		auto ned_desired_orientation = ftf::transform_orientation_enu_ned(ftf::transform_orientation_baselink_aircraft(desired_orientation));
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: fix style

File: mavros/src/plugins/setpoint_raw.cpp [hunk 5]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+		
+		body_rate.x() = 0;
+		body_rate.y() = 0;
+		body_rate.z() = -yaw_rate_scaling_*msg->yaw_rate;
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: `yaw_rate_scaling * ftf::transform_frame_yaw(msg->yaw_rate)`

File: mavros/src/plugins/setpoint_raw.cpp [hunk 6]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+		geometry_msgs::Quaternion orientation = tf::createQuaternionMsgFromRollPitchYaw( msg->roll, msg->pitch, 0);
+		
+		double thrust = std::min(1.0, std::max(0.0, msg->thrust.z * thrust_scaling_ * system_mass_kg_));
+		Eigen::Quaterniond desired_orientation;
+		Eigen::Vector3d body_rate;
+		tf::quaternionMsgToEigen(orientation, desired_orientation);
+		
+		// Transform desired orientation to represent aircraft->NED,
+		// MAVROS operates on orientation of base_link->ENU
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: remove new line

File: mavros/src/plugins/setpoint_raw.cpp [hunk 7]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+			"yaw_rate_scaling_factor or system_mass_kg parameters");
+		     	return;
+		}
+
+		//the masks are much more limited than the docs would suggest so we don't use them
+		uint8_t type_mask = 0;
+
+		geometry_msgs::Quaternion orientation = tf::createQuaternionMsgFromRollPitchYaw( msg->roll, msg->pitch, 0);
+		
+		double thrust = std::min(1.0, std::max(0.0, msg->thrust.z * thrust_scaling_ * system_mass_kg_));
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: As @mhkabir says, this should be somehow documented in a commit above this line, in order to understand how this actually works. The same thing goes for https://github.com/mavlink/mavros/pull/964/files#diff-57d2f68a3e0e5cfa2db14ac20d8845ddR110.

File: mavros/src/plugins/setpoint_raw.cpp (hunk 5 above)
Comment (Pedro-Roque) [reply mavros/src/plugins/setpoint_raw.cpp]: clarify what happends on transform_frame_yaw ?

File: mavros/src/plugins/setpoint_raw.cpp [hunk 8]
```diff
@@ -217,6 +238,47 @@ class SetpointRawPlugin : public plugin::PluginBase,
...
+	
+		if(ignore_rpyt_messages_){
+			ROS_FATAL("Recieved roll_pitch_yaw_thrust_rate message, but ignore_rpyt_messages_ is true: "
+	 	        "the most likely cause of this is a failure to specify the thrust_scaling_factor,	"
+			"yaw_rate_scaling_factor or system_mass_kg parameters");
+		     	return;
+		}
+
+		//the masks are much more limited than the docs would suggest so we don't use them
+		uint8_t type_mask = 0;
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: Why? Why not follow what is available in `setpoint_attitude`? Setting to 0 actually means that every dimension is considered, which is not actually true.

Comment (Pedro-Roque) [reply mavros/src/plugins/setpoint_raw.cpp]: This I didn't re-check. Where is masks doc?

File: mavros/src/plugins/setpoint_raw.cpp (hunk 5 above)
Comment (TSC21) [reply mavros/src/plugins/setpoint_raw.cpp]: https://github.com/mavlink/mavros/blob/master/mavros/include/mavros/frame_tf.h#L143

Comment (TSC21) [reply mavros/src/plugins/setpoint_raw.cpp]: more cleaner, and does the required job

File: mavros/src/plugins/setpoint_raw.cpp (hunk 8 above)
Comment (TSC21) [reply mavros/src/plugins/setpoint_raw.cpp]: http://mavlink.org/messages/common/#SET_POSITION_TARGET_LOCAL_NED

**Comment by Pedro-Roque** [comment]
Done :) Let me know if there's something else and I will fix it right away

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 9]
```diff
@@ -18,6 +18,7 @@
 #include <mavros/mavros_plugin.h>
 #include <mavros/setpoint_mixin.h>
 #include <eigen_conversions/eigen_msg.h>
+#include <tf/transform_datatypes.h>
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: you need to add `mav_msgs` as a dependency on `CMakeLists.txt` and `package.xml`

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 10]
```diff
@@ -26,7 +27,7 @@
 #include <geometry_msgs/PoseStamped.h>
 #include <geometry_msgs/TwistStamped.h>
 #include <mavros_msgs/Thrust.h>
-
+#include <mav_msgs/RollPitchYawrateThrust.h>
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: you need to add `mav_msgs` as a dependency on `CMakeLists.txt` and `package.xml`

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 11]
```diff
@@ -50,7 +51,9 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
 		use_quaternion(false),
 		reverse_thrust(false),
 		tf_listen(false)
-	{ }
+	{ 
+		ignore_rpyt_messages_ = false;
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: No. Put this bellow line 53 as `ignore_rpyt_messages_(false)`

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 12]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
 			send_attitude_ang_velocity(req->header.stamp, ang_vel, thrust_msg->thrust);
 	}
 
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: delete these two newlines please.

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 13]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
...
+
+	void rpyt_cb(const mav_msgs::RollPitchYawrateThrustConstPtr msg)
+	{
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: unneeded new line

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 14]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
...
+		     	return;
+		}
+
+		// Set mask to ignore everything but thrust and attitude setpoint
+		uint8_t ignore_all_except_q_and_thrust = (7 << 0);
+
+		geometry_msgs::Quaternion orientation = 
+			tf::createQuaternionMsgFromRollPitchYaw( msg->roll, msg->pitch, 0);
+		// Transforms from thrust acceleration tp a thrust force (scaling adjusts to the UAV
+		// propeller thrust, can be calculated from UAV calibration
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: typo: `... acceleration to a thrust force...`

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 15]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
...
+		tf::quaternionMsgToEigen(orientation, desired_orientation);
+		
+		// Transform desired orientation to represent aircraft->NED,
+		// MAVROS operates on orientation of base_link->ENU
+		auto ned_desired_orientation = ftf::transform_orientation_enu_ned(
+				ftf::transform_orientation_baselink_aircraft(desired_orientation));
+		
+		body_rate.x() = 0;
+		body_rate.y() = 0;
+		body_rate.z() = yaw_rate_scaling*ftf::detail::transform_frame_yaw(msg->yaw_rate);
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: `body_rate.z() = yaw_rate_scaling * ftf::detail::transform_frame_yaw(msg->yaw_rate);`

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 16]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
...
+	
+		set_attitude_target(
+			msg->header.stamp.toNSec() / 1000000,
+			ignore_all_except_q_and_thrust,
+			ned_desired_orientation,
+			body_rate,
+			thrust);
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: unneeded newline

File: mavros/src/plugins/setpoint_attitude.cpp [hunk 17]
```diff
@@ -215,6 +238,48 @@ class SetpointAttitudePlugin : public plugin::PluginBase,
...
+	
+		set_attitude_target(
+			msg->header.stamp.toNSec() / 1000000,
+			ignore_all_except_q_and_thrust,
+			ned_desired_orientation,
+			body_rate,
+			thrust);
+			
+	}
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_attitude.cpp]: unneeded newline

File: mavros/src/plugins/setpoint_raw.cpp [hunk 18]
```diff
@@ -217,6 +219,7 @@ class SetpointRawPlugin : public plugin::PluginBase,
 				body_rate,
 				req->thrust);
 	}
+
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: unneeded newline

File: mavros/src/plugins/setpoint_raw.cpp [hunk 19]
```diff
@@ -17,6 +17,8 @@
 #include <mavros/mavros_plugin.h>
 #include <mavros/setpoint_mixin.h>
 #include <eigen_conversions/eigen_msg.h>
+#include <mav_msgs/RollPitchYawrateThrust.h>
```
Comment (TSC21) [inline mavros/src/plugins/setpoint_raw.cpp]: you don't need this here anymore

**Comment by TSC21** [comment]
Jenkins test this please.

**Comment by Pedro-Roque** [comment]
On Jenkins:

"RuntimeError: Could not resolve the rosdep key 'mav_msgs'"

**Comment by TSC21** [comment]
Weird. Please try replace it with `mav_tools` and check if it happens the same way. `mav_msgs` is also a ros package so I don't understand why it fails.

**Comment by Pedro-Roque** [comment]
Yep it's wierd, on my empty workspace it compiles well with just mavros and mav_msgs packages.

**Comment by Pedro-Roque** [comment]
Made a clean pull request. Closing this one.

**Pull request closed without merging**